src/onion_client.md

```diff
@@ -0,0 +1,150 @@
+3. Set path timeouts using `{sb.friend_num, sb.path_num}`
+4. Add the announce node to `announce_list`
+5. Add the node to `path_nodes`
+6. Get nodes from the payload and ping them (if any)
+
+The process of pinging a node is the following:
+
+1. -//-
+
+A sequence chart for the beginning of self announce process (messages are sent via onion):
+
+<object data="static/announce_start.svg" type="image/svg+xml"></object>
+
+## Friend search and DHTPK announce
+
+The overall process has two steps:
+
+- First, we use announce requests to find announce nodes that stores paths to our friend
+	- This process is similar to the announce process
+- Then, we send (via onion) `DataRequest` to found announce nodes
+
+This is the chart of the second step:
+
+<object data="static/dhtpk.svg" type="image/svg+xml"></object>
+
+Specifically:
+
+1. We start with constructing `DhtPkData`:
+	- `no_replay = now` where `now` is the current time
+	- `dht_pk = dht_pk` where `dht_pk` is our dht public key
+	- `nodes = closest` where `closest` is a list of closest to us dht nodes
+2. Serialize it into bytes and send as an onion data or via dht
+
+The process of sending onion data is the following:
+
+1. The `client_list` of a friend is cleaned up of timeouted nodes
+2. Good nodes are with `is_stored != 0`
+3. It should be more than `(num_nodes - 1) / 4 + 1` good nodes to continue where `num_nodes` is the number of nodes in `client_list`
+4. Generate a random nonce
+5. The data is encrypted using friend's real public key, our real secret key and the nonce
+6. Construct `OnionData`:
+	- `real_pk` is our real public key
+	- `dht_pk_data` is the encrypted data
+7. For each good node:
+	1. Get a random friend path
+	2. Construct a `DataRequest`:
+		- `dest_pk = friend.real_pk`
+		- `nonce = nonce`. The same nonce that we used before
+		- `temp_pk = random_pk`. We generate a random keypair
+		- The `payload` is the onion data encryped with the node's data public key and the random secret key, using the same `nonce`
+	3. Send the request via the onion path
+
+## Notes on data structures
+
+Both `path_nodes` and `announce_list` are arrays of limited size. But the way they are updated are different.
+
+When `path_nodes` is full, adding a new element replaces an old one in a circular manner: first adding an element replaces `path_nodes[0]`, then `path_nodes[1]`, and so on til we get to the end of the array. After that, we begin again with `path_nodes[0]`.
+
+`announce_list` is different. It is sorted by distance to real our public key. When an element is added, it is checked against the farest node. If the element is closer to us than the node, the node is removed and the element is inserted. Otherwise, the element is simply discarded.
+
+Some packet data structures:
+
+```
+struct DhtPkData {
+    no_replay: u64,
+    dht_pk: PublicKey,
+    nodes: Vec<PackedNode>,
+}
+
+layout DhtPkData {
+    ONION_DATA_DHTPK: u8,
+    u64,
```

Comment:
what means this u64?

Reply:
An `u64` large timestamp.

Reply:
Why did you assigned the name to ONION_DATA_DHTPK but omitted the other fields?

Reply:
It is not a name, it is a constant. The first byte of the data equals to `ONION_DATA_DHTPK`.